autoload/lsp/ui/vim/documentation.vim

```diff
@@ -17,51 +23,111 @@ function! s:show_documentation(event) abort
+        for l:index in range(len(l:lines))
+            if len(l:lines[l:index]) > 0
+                let l:lines[l:index] = ' ' . l:lines[l:index] . ' '
+            endif
+        endfor
+
+        call nvim_buf_set_lines(l:buffer, 0, -1, v:false, l:lines)
+        call nvim_buf_set_option(l:buffer, 'readonly', v:true)
+        call nvim_buf_set_option(l:buffer, 'modifiable', v:false)
+        call nvim_buf_set_option(l:buffer, 'filetype', l:ft.'.lsp-hover')
+
+        if !g:lsp_documentation_float_docked
+            let l:bufferlines = nvim_buf_line_count(l:buffer)
+            let l:maxwidth = max(map(getbufline(l:buffer, 1, '$'), 'strdisplaywidth(v:val)'))
+            if g:lsp_preview_max_width > 0
+                let l:maxwidth = min([g:lsp_preview_max_width, l:maxwidth])
+            endif
+            let l:width = min([float2nr(l:width), l:maxwidth])
+            let l:height = min([float2nr(l:height), l:bufferlines])
         endif
-        if l:height <= 0
-          let l:height = 1
+        if g:lsp_preview_max_height > 0
+            let l:maxheight = g:lsp_preview_max_height
+            let l:height = min([l:height, l:maxheight])
         endif
-        let s:last_popup_id = lsp#ui#vim#output#floatingpreview([])
-        call nvim_win_set_config(s:last_popup_id, {'relative': 'win', 'anchor': l:right ? 'NW' : 'NE', 'row': l:line - 1, 'col': l:col - 1, 'height': l:height, 'width': l:width})
+
+        " Height and width must be atleast 1, otherwise error
+        let l:height = (l:height < 1 ? 1 : l:height)
+        let l:width = (l:width < 1 ? 1 : l:width)
+
+        let s:last_popup_id = nvim_open_win(l:buffer, v:false, {'relative': 'editor', 'anchor': l:anchor, 'row': l:row, 'col': l:col, 'height': l:height, 'width': l:width, 'style': 'minimal'})
+        return
     endif
 
+    " Vim
+    let l:current_win_id = win_getid()
+
+    let l:right = wincol() < winwidth(0) / 2
+    if l:right
+        let l:line = a:event['row'] + 1
+        let l:col = a:event['col'] + a:event['width'] + 1 + (a:event['scrollbar'] ? 1 : 0)
+    else
+        let l:line = a:event['row'] + 1
+        let l:col = a:event['col'] - 1
+    endif
+    let s:last_popup_id = popup_create('(no documentation available)', {'line': l:line, 'col': l:col, 'pos': l:right ? 'topleft' : 'topright', 'padding': [0, 1, 0, 1]})
     call setbufvar(winbufnr(s:last_popup_id), 'lsp_syntax_highlights', l:syntax_lines)
     call setbufvar(winbufnr(s:last_popup_id), 'lsp_do_conceal', 1)
     call lsp#ui#vim#output#setcontent(s:last_popup_id, l:lines, l:ft)
-    let [l:bufferlines, l:maxwidth] = lsp#ui#vim#output#get_size_info()
-
     call win_gotoid(l:current_win_id)
-
-    if s:use_nvim_float
-        call lsp#ui#vim#output#adjust_float_placement(l:bufferlines, l:maxwidth)
-        call nvim_win_set_config(s:last_popup_id, {'relative': 'win', 'row': l:line - 1, 'col': l:col - 1})
-    endif
 endfunction
 
 function! s:close_popup() abort
```

Comment:
I think we should add `call timer_stop(s:last_timer_id)` in here.
Because `s: show_documentation` is now debounced so I afraid to open popup after call` s: close_popup`.

Reply:
Other debounce related codes are looks good to me.

Reply:
I added it in the latest commit. Please check if correctly implemented.